Who else is using CNVkit?
=========================

`Google Scholar
<https://scholar.google.com/scholar?cites=206689808041666451&as_sdt=2005&sciodt=0,5&hl=en>`_
lists some of the studies where CNVkit has been used by other researchers.
We'd like to highlight:

* TRACERx Renal Consortium studies:

  * Turajlic, S. *et al.* (2018).
    `Tracking Cancer Evolution Reveals Constrained Routes to Metastases: TRACERx
    Renal.
    <https://doi.org/10.1016/j.cell.2018.03.057>`_
    *Cell*, 173(3), 581-594.e12
  * Turajlic, S. *et al.* (2018).
    `Deterministic Evolutionary Trajectories Influence Primary Tumor Growth:
    TRACERx Renal.
    <https://doi.org/10.1016/j.cell.2018.03.043>`_
    *Cell*, 173(3), 595-610.e11

* Seed, G. *et al.* (2017).
  `Gene copy number estimation from targeted next generation sequencing of
  prostate cancer biopsies: Analytic validation and clinical qualification.
  <http://dx.doi.org/10.1158/1078-0432.CCR-17-0972>`_
  *Clinical Cancer Research*, 23(20), 6070–6077
* McCreery, M.Q. *et al.* (2015).
  `Evolution of metastasis revealed by mutational landscapes of chemically
  induced skin cancers. <http://dx.doi.org/10.1038/nm.3979>`_
  *Nature Medicine*, 21, 1514–1520
* Shain, A.H. *et al.* (2015).
  `Exome sequencing of desmoplastic melanoma identifies recurrent NFKBIE
  promoter mutations and diverse activating mutations in the MAPK pathway.
  <http://dx.doi.org/10.1038/ng.3382>`_
  *Nature Genetics*, 47(10), 1194-1199
* Shain, A.H. *et al.* (2015).
  `The Genetic Evolution of Melanoma from Precursor Lesions.
  <http://dx.doi.org/10.1056/NEJMoa1502583>`_
  *New England Journal of Medicine*, 373(20), 1926-1936

Specific support for CNVkit is included in
`bcbio-nextgen <https://bcbio-nextgen.readthedocs.io/>`_,
`PureCN <https://github.com/lima1/PureCN>`_,
`THetA2 <http://compbio.cs.brown.edu/projects/theta/>`_,
and `MetaSV <http://bioinform.github.io/metasv/>`_.
CNVkit is also available on the commercial platforms
`DNAnexus <http://www.dnanexus.com/>`_,
`Bina RAVE <http://www.bina.com/rave>`_, and
`Diploid InHelix <http://www.diploid.com/inhelix>`_.

Finally, CNVkit can :ref:`export` files to several standard formats that can be
used with many other software packages, including `BioDiscovery Nexus Copy
Number <http://www.biodiscovery.com/nexus-copy-number/>`_ and `Integrative
Genomics Viewer (IGV) <http://software.broadinstitute.org/software/igv/>`_.
